drivers/gpio/gpio_sam4l.c

```diff
@@ -0,0 +1,305 @@
+/*
+ * Copyright (c) 2018 Justin Watson
+ * Copyright (c) 2020 Gerson Fernando Budke <nandojve@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT atmel_sam4l_gpio
+
+#include <errno.h>
+#include <kernel.h>
+#include <device.h>
+#include <init.h>
+#include <soc.h>
+#include <drivers/gpio.h>
+
+#include "gpio_utils.h"
+
+typedef void (*config_func_t)(const struct device *dev);
+
+struct gpio_sam_config {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+	Gpio *regs;
+	config_func_t config_func;
+	uint32_t periph_id;
+};
+
+struct gpio_sam_runtime {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+	sys_slist_t cb;
+};
+
+#define DEV_CFG(dev) \
+	((const struct gpio_sam_config * const)(dev)->config)
+#define DEV_DATA(dev) \
+	((struct gpio_sam_runtime * const)(dev)->data)
+
+#define GPIO_SAM_ALL_PINS    0xFFFFFFFF
+
+static int gpio_sam_port_configure(const struct device *dev,
+				   uint32_t mask,
+				   gpio_flags_t flags)
+{
+	const struct gpio_sam_config * const cfg = DEV_CFG(dev);
+	Gpio * const gpio = cfg->regs;
+
+	if (flags & GPIO_SINGLE_ENDED) {
+		/* TODO: Add support for Open Source, Open Drain mode */
+		return -ENOTSUP;
+	}
+
+	if (!(flags & (GPIO_OUTPUT | GPIO_INPUT))) {
+		/* Neither input nor output mode is selected */
+
+		/* Disable the interrupt. */
+		gpio->IERC = mask;
+		/* Disable pull-up. */
+		gpio->PUERC = mask;
+		/* Disable pull-down. */
+		gpio->PDERC = mask;
+		/* Let the GPIO control the pin (instead of a peripheral). */
+		gpio->GPERS = mask;
+		/* Disable output. */
+		gpio->ODERC = mask;
+
+		return 0;
+	}
+
+	/* Setup the pin direcion. */
+	if (flags & GPIO_OUTPUT) {
+		if (flags & GPIO_OUTPUT_INIT_HIGH) {
+			/* Set the pin. */
+			gpio->OVRS = mask;
+		}
+		if (flags & GPIO_OUTPUT_INIT_LOW) {
+			/* Clear the pin. */
+			gpio->OVRC = mask;
+		}
+		/* Enable the output */
+		gpio->ODERS = mask;
+		gpio->STERC = mask;
+	} else {
+		/* Disable the output */
+		gpio->ODERC = mask;
+		gpio->STERS = mask;
```

Comment:
Interesting, GPIO DTS bindings don't allow to control schmitt-trigger mode, only pinctrl bindings. That seems to be missing since other pin features like pull-up, pull-down can be controlled via either gpio or pinctrl bindings.

In any case this code is a bit inconsistent. The input direction of the pin is always enabled. As such there is no output only mode, there is only input or input/output. As such schmitt-trigger should be either always disabled or always enabled.

Reply:
I see, I prefer keep schmitt-trigger always enabled.